File: and-bible/app/src/main/java/net/bible/android/view/activity/page/screen/Separator.java
package net.bible.android.view.activity.page.screen;

import android.content.Context;
import android.content.res.Resources;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.LinearLayout;

import net.bible.android.BibleApplication;
import net.bible.android.activity.R;
import net.bible.android.control.page.window.Window;
import net.bible.android.control.page.window.WindowControl;
import net.bible.android.view.util.TouchDelegateView;
import net.bible.android.view.util.TouchOwner;

/**
 * @author Martin Denham [mjdenham at gmail dot com]
 * @see gnu.lgpl.License for license details.<br>
 *      The copyright to this program is held by it's author.
 */
public class Separator extends View {

	private View parentLayout;
	private Window window1;
	private Window window2;
	private int numWindows;
	
	// offset absolute points from top of layout to enable correct calculation of screen weights in layout
	private float parentStartRawPx;
	
	private boolean isPortrait = true;

	// the offset of the touch from the centre of the separator - to prevent initial jerk of separator to touch point
	private int startTouchPx;
	private float startWeight1 = 1.0f;
	private float startWeight2 = 1.0f;
	
	private int lastOffsetFromEdgePx;

	// try to prevent swamping of ui thread during splitter drag
	private long lastTouchMoveEvent;
	private final static int DRAG_TOUCH_MOVE_FREQUENCY_MILLIS = 200;
	
	private LinearLayout.LayoutParams view1LayoutParams;
	private LinearLayout.LayoutParams view2LayoutParams;

	private TouchDelegateView touchDelegateView1;
	private TouchDelegateView touchDelegateView2;

	private int SEPARATOR_WIDTH;
	private int SEPARATOR_COLOUR;
	private int SEPARATOR_DRAG_COLOUR;
	
	private TouchOwner touchOwner = TouchOwner.getInstance();
	private final WindowControl windowControl;

	private static final String TAG = "Separator";
	
	public Separator(Context context, int width, View parentLayout, Window window, Window nextWindow, int numWindows, boolean isPortrait, WindowControl windowControl) {
		super(context);
		this.windowControl = windowControl;

		Resources res = BibleApplication.getApplication().getResources();
		SEPARATOR_COLOUR = res.getColor(R.color.window_separator_colour);
		SEPARATOR_DRAG_COLOUR = res.getColor(R.color.window_separator_drag_colour);
        setBackgroundColor(SEPARATOR_COLOUR);

        touchDelegateView1 = new TouchDelegateView(context, this);
        touchDelegateView2 = new TouchDelegateView(context, this);
        SEPARATOR_WIDTH = width;
        this.parentLayout = parentLayout;
        this.window1 = window;
        this.window2 = nextWindow;
        this.numWindows = numWindows;
        this.isPortrait = isPortrait;
	}
	
	/** 
	 * Must use rawY below because this view is moving and getY would give the position relative to a moving component.
	 */
	@Override
	public boolean onTouchEvent(MotionEvent event) {
	    switch (event.getAction() & MotionEvent.ACTION_MASK)   
	    {
	    case MotionEvent.ACTION_DOWN:
	    	Log.d(TAG, " y:"+event.getRawY());
	    	touchOwner.setTouchOwner(this);
	    	windowControl.setSeparatorMoving(true);
	        setBackgroundColor(SEPARATOR_DRAG_COLOUR);
	    	
	    	int[] rawParentLocation = new int[2]; 
	    	parentLayout.getLocationOnScreen(rawParentLocation);
	    	parentStartRawPx = isPortrait? rawParentLocation[1] : rawParentLocation[0];
	    	
	        startTouchPx = isPortrait? (int)event.getRawY() : (int)event.getRawX();
	        startWeight1 = view1LayoutParams.weight; //window1.getWeight();
	        startWeight2 = view2LayoutParams.weight; //window2.getWeight();
	        break;
	    case MotionEvent.ACTION_UP:
	    case MotionEvent.ACTION_POINTER_UP:
	    	Log.d(TAG, "Up x:"+event.getX()+" y:"+event.getY());
	        setBackgroundColor(SEPARATOR_COLOUR);
	    	window1.getWindowLayout().setWeight(view1LayoutParams.weight);
	    	window2.getWindowLayout().setWeight(view2LayoutParams.weight);
	    	windowControl.setSeparatorMoving(false);
	    	touchOwner.releaseOwnership(this);
	        break;
	    case MotionEvent.ACTION_MOVE:
	    	if (System.currentTimeMillis()>lastTouchMoveEvent+DRAG_TOUCH_MOVE_FREQUENCY_MILLIS) {
		    	Log.d(TAG, "Touch move accepted");
		    	int parentDimensionPx = getParentDimensionPx();

		    	// calculate y offset in pixels from top of parent layout
		    	float offsetFromEdgePx = (isPortrait? event.getRawY() : event.getRawX());
		    	
		    	// prevent going irretrievably off bottom or right edge
	    		offsetFromEdgePx = Math.min(offsetFromEdgePx, parentDimensionPx-SEPARATOR_WIDTH);
		    	
		    	// if position has moved at least one px then redraw separator
		    	if ((int)offsetFromEdgePx != lastOffsetFromEdgePx) {
		    		int changePx = (int)offsetFromEdgePx-startTouchPx;
		    		float aveScreenSize = getAveScreenSize();
		    		float variationPercent = changePx/aveScreenSize; 
		    		
			    	// change the weights of both bible views to effectively move the separator
			    	view1LayoutParams.weight = startWeight1+variationPercent;
			    	view2LayoutParams.weight = startWeight2-variationPercent;
			    	parentLayout.requestLayout();
			    	lastOffsetFromEdgePx = (int)offsetFromEdgePx;
		    	}
		    	lastTouchMoveEvent = System.currentTimeMillis();
		    	Log.d(TAG, "Touch move finished");
	    	}
	        break;
	    }   

		return true; //super.onTouchEvent(event);
	}

	private int getAveScreenSize() {
		return getParentDimensionPx()/numWindows;
	}
	
	private int getParentDimensionPx() {
		return isPortrait? parentLayout.getHeight() : parentLayout.getWidth();
	}

	private int getStartingOffsetY() {
		return +(int)parentStartRawPx + ((getTop()+getBottom())/2);
	}
	private int getStartingOffsetX() {
		return +(int)parentStartRawPx + ((getLeft()+getRight())/2);
	}
	public void setView1LayoutParams(LinearLayout.LayoutParams view1LayoutParams) {
		this.view1LayoutParams = view1LayoutParams;
	}

	public void setView2LayoutParams(LinearLayout.LayoutParams view2LayoutParams) {
		this.view2LayoutParams = view2LayoutParams;
	}

	public TouchDelegateView getTouchDelegateView1() {
		return touchDelegateView1;
	}
	public TouchDelegateView getTouchDelegateView2() {
		return touchDelegateView2;
	}
}

File: app/src/main/java/net/bible/android/view/activity/page/screen/Separator.kt
/*
 * Copyright (c) 2020-2022 Martin Denham, Tuomas Airaksinen and the AndBible contributors.
 *
 * This file is part of AndBible: Bible Study (http://github.com/AndBible/and-bible).
 *
 * AndBible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * AndBible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with AndBible.
 * If not, see http://www.gnu.org/licenses/.
 */

package net.bible.android.view.activity.page.screen

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout

import net.bible.android.BibleApplication
import net.bible.android.activity.R
import net.bible.android.control.event.ABEventBus
import net.bible.android.control.event.window.CurrentWindowChangedEvent
import net.bible.android.control.page.window.WindowControl
import net.bible.android.view.activity.page.BibleView
import net.bible.android.view.util.TouchDelegateView
import net.bible.android.view.util.TouchOwner

/**
 * @author Martin Denham [mjdenham at gmail dot com]
 */
@SuppressLint("ViewConstructor")
class Separator(
    context: Context,
    private val separatorWidth: Int,
    private val parentLayout: View,
    val frame1: BibleFrame,
    val frame2: BibleFrame,
    internal var numWindows: Int,
    private val isSplitVertically: Boolean,
    private val windowControl: WindowControl
) : View(context)
{
    private val activeWindow get() = windowControl.windowRepository.activeWindow

    private val window1 get() = frame1.window
    private val window2 get() = frame2.window

    // offset absolute points from top of layout to enable correct calculation of screen weights in layout
    private var parentStartRawPx: Float = 0.toFloat()

    // the offset of the touch from the centre of the separator - to prevent initial jerk of separator to touch point
    private var startTouchPx: Int = 0
    private var startWeight1 = 1.0f
    private var startWeight2 = 1.0f

    private var lastOffsetFromEdgePx: Int = 0

    // try to prevent swamping of ui thread during splitter drag
    private var lastTouchMoveEvent: Long = 0

    lateinit var view1LayoutParams: LinearLayout.LayoutParams
    lateinit var view2LayoutParams: LinearLayout.LayoutParams

    val touchDelegateView1 = TouchDelegateView(context, this)
    val touchDelegateView2 = TouchDelegateView(context, this)

    private val aveScreenSize: Int
        get() = parentDimensionPx / numWindows

    private val parentDimensionPx: Int
        get() {
            val displayMetrics = Resources.getSystem().displayMetrics
            return if (isSplitVertically) displayMetrics.heightPixels else displayMetrics.widthPixels
        }

	private val res = BibleApplication.application.resources

    private val isActive get() = activeWindow.id == window1.id || activeWindow.id == window2.id

    private val separatorResource get () = if (isActive) R.drawable.separator_active else R.drawable.separator
	private val dragResource = R.drawable.separator_drag

    private fun updateBackground() {
        setBackgroundResource(separatorResource)
    }

    override fun onDetachedFromWindow() {
        ABEventBus.unregister(this)
        super.onDetachedFromWindow()
    }

    override fun onAttachedToWindow() {
        updateBackground()
        ABEventBus.register(this)
        super.onAttachedToWindow()
    }

    fun onEvent(event: CurrentWindowChangedEvent) {
        updateBackground()
    }

    fun onEvent(event: BibleView.BibleViewTouched) {
        if(event.onlyTouch) {
            release()
        }
    }

    fun release() {
        if(windowControl.isSeparatorMoving) {
            setBackgroundResource(separatorResource)
            window1.weight = view1LayoutParams.weight
            window2.weight = view2LayoutParams.weight
            windowControl.isSeparatorMoving = false
            TouchOwner.releaseOwnership()
        }
    }

    /**
     * Must use rawY below because this view is moving and getY would give the position relative to a moving component.
     */
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action and MotionEvent.ACTION_MASK) {
            MotionEvent.ACTION_DOWN -> {
                Log.i(TAG, " y:" + event.rawY)
                TouchOwner.setTouchOwner(this)
                windowControl.isSeparatorMoving = true
                setBackgroundResource(dragResource)

                val rawParentLocation = IntArray(2)
                parentLayout.getLocationOnScreen(rawParentLocation)
                parentStartRawPx = (if (isSplitVertically) rawParentLocation[1] else rawParentLocation[0]).toFloat()

                startTouchPx = if (isSplitVertically) event.rawY.toInt() else event.rawX.toInt()
                startWeight1 = view1LayoutParams.weight
                startWeight2 = view2LayoutParams.weight
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP -> {
                Log.i(TAG, "Up x:" + event.x + " y:" + event.y)
                release()
            }
            MotionEvent.ACTION_MOVE -> if (System.currentTimeMillis() > lastTouchMoveEvent + DRAG_TOUCH_MOVE_FREQUENCY_MILLIS) {
                Log.i(TAG, "Touch move accepted")
                val parentDimensionPx = parentDimensionPx

                // calculate y offset in pixels from top of parent layout
                var offsetFromEdgePx = if (isSplitVertically) event.rawY else event.rawX

                // prevent going irretrievably off bottom or right edge
                offsetFromEdgePx = Math.min(offsetFromEdgePx, (parentDimensionPx - separatorWidth).toFloat())

                // if position has moved at least one px then redraw separator
                if (offsetFromEdgePx.toInt() != lastOffsetFromEdgePx) {
                    val changePx = offsetFromEdgePx.toInt() - startTouchPx
                    val aveScreenSize = aveScreenSize.toFloat()
                    val variationPercent = changePx / aveScreenSize

                    // change the weights of both bible views to effectively move the separator
                    view1LayoutParams.weight = startWeight1 + variationPercent
                    view2LayoutParams.weight = startWeight2 - variationPercent
                    parentLayout.requestLayout()
                    lastOffsetFromEdgePx = offsetFromEdgePx.toInt()
                }
                lastTouchMoveEvent = System.currentTimeMillis()
                Log.i(TAG, "Touch move finished")
            }
        }

        return true
    }

    companion object {
        private const val DRAG_TOUCH_MOVE_FREQUENCY_MILLIS = 0

        private val TAG = "Separator"
    }
}
